sqld/src/hrana/session.rs

```diff
@@ -0,0 +1,193 @@
+                .await
+                .context("Could not create a database connection")?;
+            let stream = Stream { db };
+            session.streams.insert(stream_id, stream);
+
+            Ok(proto::Response::OpenStream(proto::OpenStreamResp {}))
+        }
+        proto::Request::CloseStream(req) => {
+            session.streams.remove(&req.stream_id);
+            Ok(proto::Response::CloseStream(proto::CloseStreamResp {}))
+        }
+        proto::Request::Execute(req) => {
+            let stream_id = req.stream_id;
+
+            let Some(stream) = session.streams.get_mut(&stream_id) else {
+                bail!(ResponseError::StreamNotFound { stream_id })
+            };
+
+            let result = execute_stmt(stream, req.stmt).await?;
+            Ok(proto::Response::Execute(proto::ExecuteResp { result }))
+        }
+    }
+}
+
+async fn execute_stmt(stream: &mut Stream, stmt: proto::Stmt) -> Result<proto::StmtResult> {
+    let query = proto_stmt_to_query(stmt)?;
+    let (query_result, _) = stream.db.execute_one(query).await?;
+    match query_result {
+        Ok(query_response) => Ok(proto_stmt_result_from_query_response(query_response)),
+        Err(error) => match proto_response_error_from_error(error) {
+            Ok(resp_error) => bail!(resp_error),
+            Err(error) => bail!(error),
+        },
+    }
+}
+
+fn proto_stmt_to_query(proto_stmt: proto::Stmt) -> Result<Query> {
+    let mut stmt_iter = Statement::parse(&proto_stmt.sql);
+    let stmt = match stmt_iter.next() {
+        Some(stmt_res) => stmt_res?,
+        None => bail!(ResponseError::SqlNoStmt),
+    };
+
+    if stmt_iter.next().is_some() {
+        bail!(ResponseError::SqlManyStmts)
+    }
+
+    let params = proto_stmt
+        .args
+        .into_iter()
+        .map(proto_value_to_value)
+        .collect();
+    let params = Params::Positional(params);
+    Ok(Query { stmt, params })
+}
+
+fn proto_stmt_result_from_query_response(query_response: QueryResponse) -> proto::StmtResult {
+    let QueryResponse::ResultSet(result_set) = query_response;
+    let proto_cols = result_set
+        .columns
+        .into_iter()
+        .map(|col| proto::Col {
+            name: Some(col.name),
+        })
+        .collect();
+    let proto_rows = result_set
+        .rows
+        .into_iter()
+        .map(|row| row.values.into_iter().map(proto_value_from_value).collect())
+        .collect();
+    proto::StmtResult {
+        cols: proto_cols,
+        rows: proto_rows,
+        affected_row_count: result_set.affected_row_count,
+    }
+}
+
+fn proto_value_to_value(proto_value: proto::Value) -> Value {
+    match proto_value {
+        proto::Value::Null => Value::Null,
+        proto::Value::Integer { value } => Value::Integer(value),
+        proto::Value::Float { value } => Value::Real(value),
+        proto::Value::Text { value } => Value::Text(value),
+        proto::Value::Blob { value } => Value::Blob(value),
+    }
+}
```

Comment:
would implementing `From` here make sense?

Reply:
I strongly prefer explicit functions to overloading `From`. Overloads make it harder to follow what is going on, especially when we are juggling multiple types with very similar names (such as `Value` and `proto::Value` here). `proto_value_to_value(value)` is much more descriptive than `value.into()`.

(In general, I tend to optimize for simplicity, and a plain function is simpler than an overloaded one.)

Reply:
I understand your POV, yet I also like to optimize for the principle of least astonishment. When you want to know if some type can be converted to another type in Rust, the usual thing you go look for is a `From` implementation. If you don't find it, you'll just assume that it doesn't exist and this creates confusion. Moreover, `From` creates an explicit link between two types (since the other type gets the reciprocal `Into` implementation for free).

I tend to agree that using `into()` makes it somewhat harder to follow what's going on, but its also because it usually don't really matter, but I find it hard to argue that `SomeType::from(something)` affects comprehension in any ways?

Reply:
Ok, you are here at home, let's use a trait :)